### ISSUE TITLE
Memory leak when computing the hessian of a jitted function

### ISSUE THREAD

**Issue by harwiltz** [google/jax#7621]
Please:

- [x] Check for duplicate issues.
- [x] Provide a complete example of how to reproduce the bug, wrapped in triple backticks like this:

```python
import jax
import jax.numpy as jnp

f = lambda x: jnp.square(x).mean()
jf = jax.jit(f)

x = jax.random.uniform(jax.random.PRNGKey(0), shape=(8, 4))

# Start monitoring memory with htop
while True:
    try:
        # let this run for a bit, notice memory usage doesn't increase
        y = jax.hessian(f)(x)
    except KeyboardInterrupt:
        pass

while True:
    # let this run, notice memory usage starts increasing immediately
    # also, the memory usage doesn't appear to stop increasing ever
    y = jax.hessian(jf)(x)
```
Also, note that the equivalent code with `jax.grad` (or `jax.jacfwd` for multidimensional functions) replacing `jax.hessian` *does not* cause a memory leak.

As a result of this leak, my python process ends up getting killed by my OS (due to OOM), so I don't ever see any python or jax traces.

**Comment by mattjj** [comment]
Thanks for the question!

At first I thought could be an artifact of [asynchronous dispatch](https://jax.readthedocs.io/en/latest/async_dispatch.html). (The `jit` dispatch path is different from the non-`jit` dispatch path, so that may play a role in why `f` and `jf` behave differently.)

But I ran this version, which forces synchronizations, and it seems like memory is still growing without bound (on the CPU backend):

```python
while True:
    jax.hessian(jf)(x).block_until_ready()
```

I then tried running the script like this:

```
env JAX_LOG_COMPILES=1 python issue7621.py
```

and saw that we're doing lots of recompiles here! Maybe that should've been obvious from your clue about `grad`/`jacfwd` behaving differently.

Somehow `hessian` is causing compilation cache misses here...

**Comment by harwiltz** [comment]
@mattjj Nice catch, I wasn't aware of the `JAX_LOG_COMPILES` variable. I wonder if declaring the function `h = hessian(jf)` outside the loop and calling `h(x)` within the loop would still cause the issue. I'll give it a try (though admittedly I don't know much about how `jit` works, so this may not make sense ;) )

**Comment by mattjj** [comment]
This _should_ be getting cache hits, so it's definitely a bug. The situation should be basically the same as for `grad`, so there's some difference between `grad` and `hessian` which we need to pin down...

**Comment by mattjj** [comment]
I see the issue. It's a problem with `vmap`. It's a bit hard to unpack in full detail, so for now I'm just going to leave some notes for myself and other jax devs: in [this cache key](https://github.com/google/jax/blob/a3387aeb0c26e8334d874160a00db7d5e174264b/jax/linear_util.py#L255), on each invocation we're getting different values because `key[0][7][1][0].payload` is a different [temporary axis name](https://github.com/google/jax/blob/a3387aeb0c26e8334d874160a00db7d5e174264b/jax/_src/api.py#L1259) each time. That's happening in turn because `jacrev's` [`pullback`](https://github.com/google/jax/blob/a3387aeb0c26e8334d874160a00db7d5e174264b/jax/_src/api.py#L1033-L1035), which gets vmapped, is a fresh Python callable each time.

**Comment by mattjj** [comment]
More notes to self:

We added temporary axis names in #917 for soft_pmap (now replaced by xmap). But the issue here is about vmap using temporary axis names.

The reason we needed them for soft_pmap/xmap is that those implementations split axes, and so they need to generate new names which could be used in collectives (and hence have to be distinguished from one another, etc).

But here with vmap, when we aren't given a name, the axis can't be involved in any collectives. So it seems like we should just leave it as None. Unless I'm forgetting something...

**Comment by mattjj** [comment]
Adam and I added this TempAxisName stuff to vmap in #4005. I wonder if that was just a bug...

**Comment by harwiltz** [comment]
@mattjj cool stuff. Thanks! When my schedule frees up I'm going to learn more about the inner workings of jax, hopefully I can be of more help next time.

**Comment by mattjj** [comment]
[Autodidax](https://jax.readthedocs.io/en/latest/autodidax.html) is a good resource for that! Though because it doesn't (yet) have named axes and collectives, it doesn't have anything related to this particular issue.

**Comment by patrick-kidger** [comment]
I just wanted to check if I may have hit another instance of this bug, or if I've hit something else altogether.

This seems to manifest as a three-way interaction between `jit`, `vmap` and `jvp`, and moreover (probably a separate bug) exhibits different behaviour depending on whether you use `def` or a `lambda` (!)

```python
import jax
import jax.numpy as jnp

@jax.jit
def f(x):
    print("called")

class M:
    def __call__(self, x):
        f(x)

jax.tree_util.register_pytree_node(M, lambda m: ((), None), lambda _, __: M())

def g(m, x):
    jax.vmap(m)(x)

m = M()
x = jnp.zeros(2)
jax.jvp(g, (m, x), (m, x))  # prints "called"
jax.jvp(g, (m, x), (m, x))  # prints "called" -- the JIT'd function has been compiled twice!
```

The error goes away if any of:
- The `jvp` is removed and `g` called directly: `g(m, x)`.
- The `vmap` is removed and `m` called directly: `m(x)`.
- An `axis_name` is passed to `vmap`;
- `jax.vmap(m)(x)` is replaced with
```python
def wrap(m, x):
    return m(x)
jax.vmap(wrap, in_axes=(None, 0))(m, x)`
```

But remarkably it does **not** go away if the same line is replaced with `jax.vmap(lambda m, x: m(x), in_axes=(None, 0))(m, x)`!

I think there may be as many as three errors here:
- Unnecessary recompilation;
- That the `def wrap` version is treated differently to the original version;
- That the `lambda` version is treated differently to the `def wrap` version.

WDYT? (I don't think I'm doing anything wrong as an end-user?)
I can open a separate issue for these if you think they're separate to this issue; I'm not sufficiently familiar with JAX internals to say.

**Comment by mattjj** [comment]
I think it's an instance of the same bug. Running your code against current HEAD, I only see 'called' printed once.
